include/ruby/random.h

```diff
@@ -0,0 +1,70 @@
+#ifndef RUBY_RANDOM_H
+#define RUBY_RANDOM_H 1
+/**
+ * @file
+ * @author     $Author$
+ * @date       Sat May  7 11:51:14 JST 2016
+ * @copyright  2007-2020 Yukihiro Matsumoto
+ * @copyright  This  file  is   a  part  of  the   programming  language  Ruby.
+ *             Permission  is hereby  granted,  to  either redistribute  and/or
+ *             modify this file, provided that  the conditions mentioned in the
+ *             file COPYING are met.  Consult the file for details.
+ */
+
+#include "ruby/ruby.h"
+
+#if defined(__cplusplus)
+extern "C" {
+#if 0
+} /* satisfy cc-mode */
+#endif
+#endif
+
+RUBY_SYMBOL_EXPORT_BEGIN
+
+typedef struct {
+    VALUE seed;
+} rb_random_t;
+
+typedef void rb_random_init_func(rb_random_t *, const uint32_t *, size_t);
+typedef unsigned int rb_random_get_int32_func(rb_random_t *);
+typedef void rb_random_get_bytes_func(rb_random_t *, void *, size_t);
+
+typedef struct {
+    size_t default_seed_bits;
+    rb_random_init_func *init;
+    rb_random_get_int32_func *get_int32;
+    rb_random_get_bytes_func *get_bytes;
+} rb_random_interface_t;
```

Comment:
Some random generators,  such as dSFMT, can generate real random numbers directly.
How to support directly generation of real random numbers in such generators?

Reply:
What signature does that function have?

Reply:
The signature of the function that generates a real random number in `[0, 1)` is the same as `rb_random_get_int32_func`.

In the original implementation of dSFMT, the prototype is:

```c
double dsfmt_genrand_close_open(dsfmt_t *dsfmt);
```

You can find this line at https://github.com/MersenneTwister-Lab/dSFMT/blob/master/dSFMT.h#L206

Reply:
We can make `get_int32` is required, and `get_double` is optional.
If a given generator type provides `get_double` function, `random_real` can use `get_double`.

Reply:
We need not only close-open range but also close-close range for inclusive `Range`.